### PR TITLE
[com_categories] - convert to prepared statement

### PR DESCRIPTION
**Pull request by alikon**
### Summary of Changes
use prepared statement for SQL

### Testing Instructions
test com_categories

### Expected result
works as before

### Actual result
N/A

### CODE REVIEW

File: administrator/components/com_categories/Field/CategoryeditField.php [hunk 1]
```diff
@@ -176,61 +177,73 @@ protected function getOptions()
...
+				$language = explode(',', $this->element['language']);
 			}
 			else
 			{
-				$language = $db->quote($this->element['language']);
+				$language = $this->element['language'];
 			}
 
-			$query->where($db->quoteName('a.language') . ' IN (' . $language . ')');
+			$query->whereIn($db->quoteName('a.language'), $language);
```
Comment (Quy) [inline administrator/components/com_categories/Field/CategoryeditField.php]:
```suggestion
			$query->whereIn($db->quoteName('a.language'), $language, ParameterType::STRING);
```

File: administrator/components/com_categories/Field/Modal/CategoryField.php [hunk 2]
```diff
@@ -116,10 +117,12 @@ function jSelectCategory_" . $this->id . "(id, title, object) {
 		if ($value)
 		{
 			$db    = Factory::getDbo();
+			$value = (int) $value;
```
Comment (Quy) [inline administrator/components/com_categories/Field/Modal/CategoryField.php]:
```suggestion
```
Already declared on line 67.

File: administrator/components/com_categories/Model/CategoriesModel.php [hunk 3]
```diff
@@ -167,70 +168,107 @@ protected function getListQuery()
...
-			$query->where('a.extension = ' . $db->quote($extension));
+			$query->where($db->quoteName('a.extension') . ' = :extension')
+				->bind(':extension', $extension);
 		}
 
 		// Filter on the level.
 		if ($level = $this->getState('filter.level'))
 		{
-			$query->where('a.level <= ' . (int) $level);
+			$level = (int) $level;
```
Comment (Quy) [inline administrator/components/com_categories/Model/CategoriesModel.php]: Apply `(int)` to line 238 and remove this.

File: administrator/components/com_categories/Model/CategoriesModel.php [hunk 4]
```diff
@@ -167,70 +168,107 @@ protected function getListQuery()
...
+			$query->where($db->quoteName('a.level') . ' <= :level')
+				->bind(':level', $level, ParameterType::INTEGER);
 		}
 
 		// Filter by access level.
 		if ($access = $this->getState('filter.access'))
 		{
-			$query->where('a.access = ' . (int) $access);
+			$access = (int) $access;
```
Comment (Quy) [inline administrator/components/com_categories/Model/CategoriesModel.php]: Apply (int) to line 246 and remove this.

File: administrator/components/com_categories/Model/CategoriesModel.php [hunk 5]
```diff
@@ -240,32 +278,50 @@ protected function getListQuery()
...
 			}
 			else
 			{
-				$search = $db->quote('%' . str_replace(' ', '%', $db->escape(trim($search), true) . '%'));
-				$query->where('(a.title LIKE ' . $search . ' OR a.alias LIKE ' . $search . ' OR a.note LIKE ' . $search . ')');
+				$search = '%' . str_replace(' ', '%', trim($search)) . '%';
+				$query->extendWhere(
+					'AND',
+					[
+						$db->quoteName('a.title') . ' LIKE :title ',
```
Comment (Quy) [inline administrator/components/com_categories/Model/CategoriesModel.php]:
```suggestion
						$db->quoteName('a.title') . ' LIKE :title',
```

File: administrator/components/com_categories/Model/CategoriesModel.php [hunk 6]
```diff
@@ -240,32 +278,50 @@ protected function getListQuery()
...
+						$db->quoteName('a.alias') . ' LIKE :alias ',
```
Comment (Quy) [inline administrator/components/com_categories/Model/CategoriesModel.php]:
```suggestion
						$db->quoteName('a.alias') . ' LIKE :alias',
```

File: administrator/components/com_categories/Model/CategoriesModel.php [hunk 7]
```diff
@@ -240,32 +278,50 @@ protected function getListQuery()
...
+						$db->quoteName('a.note') . ' LIKE :note ',
```
Comment (Quy) [inline administrator/components/com_categories/Model/CategoriesModel.php]:
```suggestion
						$db->quoteName('a.note') . ' LIKE :note',
```

File: administrator/components/com_categories/Model/CategoryModel.php [hunk 8]
```diff
@@ -632,28 +633,34 @@ public function save($data)
...
 			// Get associationskey for edited item
 			$db    = $this->getDbo();
+			$id    = (int) $table->id;
 			$query = $db->getQuery(true)
 				->select($db->quoteName('key'))
 				->from($db->quoteName('#__associations'))
-				->where($db->quoteName('context') . ' = ' . $db->quote($this->associationsContext))
-				->where($db->quoteName('id') . ' = ' . (int) $table->id);
+				->where($db->quoteName('context') . ' = :associationscontext')
+				->bind('associationscontext', $this->associationsContext)
```
Comment (Quy) [inline administrator/components/com_categories/Model/CategoryModel.php]: Swap lines.

File: administrator/components/com_categories/Model/CategoryModel.php [hunk 9]
```diff
@@ -680,11 +687,15 @@ public function save($data)
...
 				$key = md5(json_encode($associations));
 				$query->clear()
-					->insert('#__associations');
+					->insert($db->quoteName('#__associations'));
 
 				foreach ($associations as $id)
 				{
-					$query->values(((int) $id) . ',' . $db->quote($this->associationsContext) . ',' . $db->quote($key));
+					$id = (int) $id;
+					$query->values(':id$id , :associationscontext$id, :key$id')
```
Comment (Quy) [inline administrator/components/com_categories/Model/CategoryModel.php]:
```suggestion
					$query->values(':id$id, :associationscontext$id, :key$id')
```

File: administrator/components/com_categories/Model/CategoriesModel.php (hunk 4 above)
Comment (alikon) [reply administrator/components/com_categories/Model/CategoriesModel.php]: done

File: administrator/components/com_categories/Model/CategoriesModel.php (hunk 3 above)
Comment (alikon) [reply administrator/components/com_categories/Model/CategoriesModel.php]: done

File: administrator/components/com_categories/Model/CategoryModel.php (hunk 8 above)
Comment (alikon) [reply administrator/components/com_categories/Model/CategoryModel.php]: done

**Comment by Quy** [comment]
> mysqli_stmt::bind_param(): Number of variables doesn't match number of parameters in prepared statement in \libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 430

Edit a category but unable to save.

File: administrator/components/com_categories/Model/CategoryModel.php [hunk 10]
```diff
@@ -1052,11 +1070,14 @@ protected function batchCopy($value, $pks, $contexts)
 			// Copy rules
 			$query->clear()
 				->update($db->quoteName('#__assets', 't'))
-				->join('INNER', $db->quoteName('#__assets', 's') .
-					' ON ' . $db->quoteName('s.id') . ' = ' . $oldAssetId
+				->join('INNER',
+					$db->quoteName('#__assets', 's'),
+					$db->quoteName('s.id') . ' = :oldid',
```
Comment (Quy) [inline administrator/components/com_categories/Model/CategoryModel.php]:
```suggestion
					$db->quoteName('s.id') . ' = :oldid'
```

**Comment by alikon** [comment]
that i could be burned on  :fire: 
:smile: 
on postgresql a save/new category works
on pdo mysql,  don't, like has been correctly reported 
:scream: 
the earth is becoming to be flat ....

File: administrator/components/com_categories/Model/CategoryModel.php [hunk 11]
```diff
@@ -632,28 +633,34 @@ public function save($data)
...
-				);
+				$query->where('(' . $db->quoteName('id') . ' IN (' . implode(',', $query->bindArray($associations)) . ') OR '
+					. $db->quoteName('key') . ' = :oldkey)'
+				)
+					->bind(':oldkey', $oldKey);
 			}
 			else
 			{
-				$query->where($db->quoteName('key') . ' = ' . $db->quote($oldKey));
+				$query->where($db->quoteName('key') . ' = :key')
```
Comment (Quy) [inline administrator/components/com_categories/Model/CategoryModel.php]: The culprit is here, but I don't understand why. `$oldKey` is `NULL`.

Comment (SharkyKZ) [reply administrator/components/com_categories/Model/CategoryModel.php]: Binding `null` values is not supported yet. Need to update database package (pending PR https://github.com/joomla/joomla-cms/pull/27044).

@alikon you can write these queries like here https://github.com/joomla/joomla-cms/blob/3ae248450652196d7f341375410539ac9af3aafc/administrator/components/com_menus/Model/ItemModel.php#L1606-L1655

Comment (Quy) [reply administrator/components/com_categories/Model/CategoryModel.php]: @alikon #27044 fixes this issue. Maybe you can test it so you don't have to rewrite these queries.

Comment (SharkyKZ) [reply administrator/components/com_categories/Model/CategoryModel.php]: I still suggest rewriting this part. We don't need to run the query in some cases.

Comment (alikon) [reply administrator/components/com_categories/Model/CategoryModel.php]: yeah i should test #27044 for sure.... anyway in the meantime is better to rewrite this part...

Comment (alikon) [reply administrator/components/com_categories/Model/CategoryModel.php]: luckily  #27044 already merged....
p.s
queries rewrited.... please retest

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 8ad38830758ec04c352e7875be6a85a24fb5c196<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27205">issues.joomla.org/tracker/joomla-cms/27205</a>.</sub>

File: administrator/components/com_categories/Field/CategoryeditField.php [hunk 12]
```diff
@@ -332,10 +345,25 @@ protected function getOptions()
 			}
 		}
 
-		if (($this->element['parent'] == true || $jinput->get('option') == 'com_categories')
-			&& (isset($row) && !isset($options[0]))
+		if ($oldCat != 0 && ($this->element['parent'] == true || $jinput->get('option') == 'com_categories')
```
Comment (SharkyKZ) [inline administrator/components/com_categories/Field/CategoryeditField.php:348]: Why the changes here and to the query below?

**Comment by jwaisner** [comment]
I have tested this item :white_check_mark: successfully on 8ad38830758ec04c352e7875be6a85a24fb5c196<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27205">issues.joomla.org/tracker/joomla-cms/27205</a>.</sub>

**Comment by SharkyKZ** [comment]
RTC.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27205">issues.joomla.org/tracker/joomla-cms/27205</a>.</sub>

**Comment by HLeithner** [comment]
Thanks
